### PR TITLE
add check for local volume option

### PR DESCRIPTION
**Pull request by moypray**
**- What I did**

Add check for creating local volume

Fix issue:
```
root@localhost:~# docker volume create -d local --name test -o type=tmpfs
test
root@localhost:~# docker run -tid -v test:/tmp ubuntu
docker: Error response from daemon: missing device in volume options.
See 'docker run --help'.
root@localhost:~#
```

I think it should fail at creating it.
```
root@localhost:~# docker volume create -d local --name test2 -o type=tmpfs
Error response from daemon: create test2: option device must be specified
root@localhost:~#
```

**- How I did it**

add check flag when creating local volume

**- How to verify it**

docker volume create -d local --name test -o type=tmpfs
It should fail to create local volume when provide wrong options.

**- Description for the changelog**

Signed-off-by: Wentao Zhang <zhangwentao234@huawei.com>

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
@moypray there's a test failing;

```
08:45:34 ----------------------------------------------------------------------
08:45:34 FAIL: docker_api_containers_test.go:1549: DockerSuite.TestContainersAPICreateMountsValidation
08:45:34 
08:45:34 case 0
08:45:34 case 1
08:45:34 case 2
08:45:34 case 3
08:45:34 case 4
08:45:34 case 5
08:45:34 case 6
08:45:34 case 7
08:45:34 case 8
08:45:34 case 9
08:45:34 docker_api_containers_test.go:1735:
08:45:34     c.Assert(status, checker.Equals, x.status, check.Commentf("%s\n%v", string(b), cases[i].config))
08:45:34 ... obtained int = 500
08:45:34 ... expected int = 201
08:45:34 ... {"message":"create hello3: option device must be specified"}
08:45:34 
08:45:34 {busybox {[{volume hello3 /foo false  <nil> 0xc4200c4540 <nil>}]}}
08:45:34 
```

**Comment by moypray** [comment]
@thaJeztah updated, thanks, wait for CI.

BTW, I supposed that we support `tmpfs | nfs` for local driver, and the `device` and `type` options are both mandatory for them.

Correct me if something is wrong.

**Comment by dsheets** [comment]
I don't think a device is necessary for bind mounts.

**Comment by moypray** [comment]
@dsheets yes, for bind mount, it is not necessary

But I wonder if we use bind mount in local volume? And the `device` option is checked here：
https://github.com/moby/moby/blob/master/volume/local/local_unix.go#L75

So I supposed that local volume just supports `nfs | tmpfs | local directory` as I said. But I am not sure.
`device` and `type` are both mandatory for `nfs` and `tmpfs`.

**Comment by moypray** [comment]
ping @thaJeztah  @cpuguy83

**Comment by dsheets** [comment]
@moypray I was confused and thinking of the `type=` key which is unnecessary for bind mounts. `device=` is, of course, how you specify the source of the bind mount. Sorry!

With that said, use of local volume bind mounts should definitely be considered when changing `local` but requiring `device=` is fine for them.

The only file system that I know of that doesn't require a device is `proc` but a dummy can be provided. As you point out, at mount time `device` is already checked for all `local` volumes.

**Comment by moypray** [comment]
@dsheets That's fine, I think we can just set `device` as mandatory key. I will rework my PR.

**Comment by moypray** [comment]
BTW, how could we use bind mount in local volume driver??

**Comment by dsheets** [comment]
`docker volume create -o o=bind -o device=/etc my_etc_volume`

**Comment by moypray** [comment]
@dsheets Got it, thanks.

File: volume/local/local_unix.go [hunk 1]
```diff
@@ -27,6 +27,9 @@ var (
 		"o":      true, // generic mount options
 		"device": true, // device to mount from
 	}
+	mandatoryKey = map[string]struct{}{
+		"device": {},
```
Comment (thaJeztah) [inline volume/local/local_unix.go]: Discussing with @cpuguy83 and we should add `type` to this list as well; can you add that?

**Comment by thaJeztah** [comment]
Generally looks good, but needs a rebase, and a check for `type`; could you update the PR @moypray ?

File: volume/local/local.go [hunk 2]
```diff
@@ -367,6 +367,13 @@ func validateOpts(opts map[string]string) error {
 			return validationError{fmt.Errorf("invalid option key: %q", opt)}
 		}
 	}
+	if len(opts) > 0 {
+		for mKey := range mandatoryKey {
+			if _, ok := opts[mKey]; !ok {
+				return validationError{fmt.Errorf("option %s must be specified", mKey)}
```
Comment (cpuguy83) [inline volume/local/local.go]: Can we use `errors.Errorf`? This passes the error with the stack trace for debugging purposes.

Maybe we can re-word this, something like "missing required option %s".

File: volume/local/local_unix.go (hunk 1 above)
Comment (cpuguy83) [inline volume/local/local_unix.go]: This needs to add "type".

**Review by cpuguy83 — CHANGES_REQUESTED** [review]

**Comment by cpuguy83** [comment]
Any news @moypray ?

**Comment by moypray** [comment]
@cpuguy83 
sorry, I do not work on docker for some time. Would some one to cherry this PR ?

**Review by vdemeester — COMMENTED** [review]
SGTM :angel: 
cc @thaJeztah @cpuguy83 I rebased it against master and took care of your comments :angel: :wink:

**Comment by cpuguy83** [comment]
```
09:24:32 volume/local/local.go:375:12:warning: invalid composite literal type validationError (gosimple)
09:24:32 volume/local/local.go:375:12:warning: invalid composite literal type validationError (interfacer)
09:24:32 volume/local/local.go:375:12:warning: invalid composite literal type validationError (unconvert)
```

**Comment by vdemeester** [comment]
re-rebased :sweat_smile:

**Review by thaJeztah — APPROVED** [review]
LGTM

**Comment by thaJeztah** [comment]
ping @cpuguy83 PTAL

**Comment by cpuguy83** [comment]
```
09:52:39 FAIL: docker_api_containers_test.go:1662: DockerSuite.TestContainersAPICreateMountsValidation
09:52:39 
09:52:39 case 0
09:52:39 case 1
09:52:39 case 2
09:52:39 case 3
09:52:39 case 4
09:52:39 case 5
09:52:39 case 6
09:52:39 case 7
09:52:39 case 8
09:52:39 case 9
09:52:39 docker_api_containers_test.go:1878:
09:52:39     c.Assert(err.Error(), checker.Contains, x.msg, check.Commentf("%v", cases[i].config))
09:52:39 ... obtained string = "Error response from daemon: create hello3: missing required option device"
09:52:39 ... substring string = "device must be specified"
09:52:39 ... {   false false false map[] false false false [] [] <nil> false busybox map[]  [] false  [] map[]  <nil> []}
09:52:39 
09:52:40 
09:52:40 ----------------------------------------------------------------------
```

**Comment by olljanat** [comment]
Derek add label: rebuild/*

**Comment by thaJeztah** [comment]
Rebased, fixed the failing test, added a missing test-case for "type", and made some small tweaks

@cpuguy83 @vdemeester PTAL

**Pull request merged**